### PR TITLE
SBP add shopping cart model

### PR DESCRIPTION
**Pull request by petro2061**
In this task added new controller for management shopping cart for the currently authorized user. Also adden model that representet product in shopping cart. Added required Dto classes, repository layer with custom methods and servise layer with the necessary methods. That methods helps add new product (in our case book) to shopping cart, update data about product and delete product from shoppind cart. This methods use in controller layer.

### CODE REVIEW

File: src/main/java/project/onlinebookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    public ShoppingCartDto getShoppingCart(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService.getShoppingCart(user.getId());
+    }
+
+    @Operation(summary = "Add new product to shopping cart",
+            description = "Allows get all new cart item to shopping cart")
+    @PreAuthorize("hasRole('USER')")
+    @PostMapping
+    public CartItemDto addCartItemToShoppingCart(
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto addCartItemToShoppingCart(
```

File: src/main/java/project/onlinebookstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            CreateCartItemRequestDto cartItemRequestDto) {
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService
+                .addCartItemToShoppingCart(user.getId(), cartItemRequestDto);
+    }
+
+    @Operation(summary = "Update product in shopping cart",
+            description = "Allows update product in the shopping cart by identifier")
+    @PutMapping("/items/{cartItemId}")
+    public CartItemDto updateCartItemInShoppingCart(
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto updateCartItemInShoppingCart(
```

File: src/main/java/project/onlinebookstore/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService
+                .updateCartItemInShoppingCart(user.getId(), cartItemId, updateRequestDto);
+    }
+
+    @Operation(summary = "Delete product in shopping cart",
+            description = "Allows delete product in the shopping cart by identifier")
+    @DeleteMapping("/items/{cartItemId}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public CartItemDto removeCartItemFromShoppingCart(
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
    public void removeCartItemFromShoppingCart(
```

File: src/main/java/project/onlinebookstore/dto/cartitem/CreateCartItemRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package project.onlinebookstore.dto.cartitem;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+
+public record CreateCartItemRequestDto(
+        @NotNull (message = "Field bookId can't be null")
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/dto/cartitem/CreateCartItemRequestDto.java]:
```suggestion
        @NotNull(message = "Field bookId can't be null")
        @Positive(
```

File: src/main/java/project/onlinebookstore/dto/cartitem/CreateCartItemRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+        Long bookId,
+        @Min(value = 1, message = "Quantity can't be less one")
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/dto/cartitem/CreateCartItemRequestDto.java]: It's better to use `@Positive`

File: src/main/java/project/onlinebookstore/dto/cartitem/CreateCartItemUpdateRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package project.onlinebookstore.dto.cartitem;
+
+import jakarta.validation.constraints.Min;
+
+public record CreateCartItemUpdateRequestDto(
+        @Min(value = 1, message = "Quantity can't be less one")
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/dto/cartitem/CreateCartItemUpdateRequestDto.java]: `@Positive`

File: src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+
+    @Override
+    @Transactional
+    public ShoppingCartDto getShoppingCart(Long userId) {
+        ShoppingCart shoppingCar = getShoppingCartForUser(userId);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java]:
```suggestion
        return shoppingCartMapper.toDto(shoppingCartRepository.findByUserId(userId));
```

File: src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+@Service
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+
+    @Override
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java]:
```suggestion
```

File: src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+        return shoppingCartDto;
+    }
+
+    @Override
+    @Transactional
+    public CartItemDto addCartItemToShoppingCart(
+            Long userId,
+            CreateCartItemRequestDto cartItemRequestDto) {
+        ShoppingCart shoppingCar = getShoppingCartForUser(userId);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart shoppingCart = getShoppingCartForUser(userId);
```

File: src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    @Override
+    @Transactional
+    public CartItemDto updateCartItemInShoppingCart(
+            Long userId,
+            Long cartItemId,
+            CreateCartItemUpdateRequestDto updateRequestDto) {
+        ShoppingCart shoppingCart = getShoppingCartForUser(userId);
+        CartItem cartItem = getCartItemById(cartItemId);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java]: Or 
```suggestion
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())
                .map(item -> {
                    item.setQuantity(itemDto.quantity());
                    return item;
                }).orElseThrow(() -> new EntityNotFoundException(
                        String.format("No cart item with id: %d for user: %d", cartItemId, userId)
                ));
```

File: src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -39,6 +44,13 @@ public UserResponseDto register(UserRegistrationRequestDto registrationRequestDt
         User user = userMapper.toUserModel(registrationRequestDto);
         user.setPassword(passwordEncoder.encode(registrationRequestDto.getPassword()));
         user.setRoles(Set.of(userRole));
-        return userMapper.toUserDto(userRepository.save(user));
+
+        User savedUser = userRepository.save(user);
+
+        ShoppingCart shoppingCart = new ShoppingCart();
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java]: create method `createShoppingCart` in ShopCartService for these logic

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Comment by petro2061** [comment]
Changed the implementation of methods at the service level. Fixed code style issues. Added new methods to service layer and storage layer.

File: src/main/java/project/onlinebookstore/controller/ShoppingCartController.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+@RequestMapping("/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @Operation(summary = "Get shopping cart",
+            description = "Returns the shopping cart with all items for the authorized user")
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping
+    public ShoppingCartDto getShoppingCart(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
```
Comment (okuzan) [inline src/main/java/project/onlinebookstore/controller/ShoppingCartController.java]: extract repeated logic to a private method

File: src/main/java/project/onlinebookstore/dto/cartitem/CreateCartItemRequestDto.java [hunk 13]
```diff
@@ -0,0 +1,12 @@
+package project.onlinebookstore.dto.cartitem;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+
+public record CreateCartItemRequestDto(
+        @NotNull(message = "Field bookId can't be null")
```
Comment (okuzan) [inline src/main/java/project/onlinebookstore/dto/cartitem/CreateCartItemRequestDto.java]: You can use Positive here too

File: src/main/java/project/onlinebookstore/model/ShoppingCart.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Setter
+@Getter
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id =?")
+@SQLRestriction(value = "is_deleted = false")
```
Comment (okuzan) [inline src/main/java/project/onlinebookstore/model/ShoppingCart.java]:
```suggestion
@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = TRUE WHERE id =?")
@SQLRestriction(value = "is_deleted = FALSE")
```

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 15]
```diff
@@ -14,4 +14,8 @@ databaseChangeLog:
   - include:
       file: db/changelog/changes/07-create-categories-table.yaml
   - include:
-      file: db/changelog/changes/08-create-books-categories-table.yaml
\ No newline at end of file
+      file: db/changelog/changes/08-create-books-categories-table.yaml
+  - include:
+      file: db/changelog/changes/09-create-shopping-carts-table.yaml
+  - include:
+      file: db/changelog/changes/10-create-cart-items-table.yaml
```
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
      file: db/changelog/changes/10-create-cart-items-table.yaml

```

File: src/main/resources/db/changelog/changes/10-create-cart-items-table.yaml [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+            baseTableName: cart_items
+            constraintName: fk_cart_item_shopping_cart
+            referencedColumnNames: id
+            referencedTableName: shopping_carts
+        - addForeignKeyConstraint:
+            baseColumnNames: book_id
+            baseTableName: cart_items
+            constraintName: fk_cart_item_book
+            referencedColumnNames: id
+            referencedTableName: books
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/10-create-cart-items-table.yaml]:
```suggestion
            referencedTableName: books

```
image
always review your PR, if you see this red sign, it means you need to add one more empty line

File: src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+        shoppingCart.getCartItems().remove(cartItem);
+        cartItem.setShoppingCart(shoppingCart);
+
+        cartItemRepository.delete(cartItem);
+    }
+
+    public void createShoppingCart(User user) {
```
Comment (okuzan) [inline src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java:102]: Add Override

**Review by okuzan — CHANGES_REQUESTED** [review]

**Comment by petro2061** [comment]
Added new private method to ShoppingCart Controller. Fixsed problem with codestyle.

File: src/main/java/project/onlinebookstore/controller/ShoppingCartController.java [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+            @Parameter(description = "Represents the cart item identifier")
+            @PathVariable @Positive Long cartItemId
+    ) {
+        shoppingCartService
+                .removeCartItemFromShoppingCart(
+                        getAuthenticationUser(authentication).getId(),
+                        cartItemId);
+    }
+
+    private User getAuthenticationUser(Authentication authentication) {
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
    private Long getAuthenticationUser(Authentication authentication) {
```

return Long userId

File: src/main/java/project/onlinebookstore/dto/cartitem/CreateCartItemRequestDto.java [hunk 19]
```diff
@@ -0,0 +1,11 @@
+package project.onlinebookstore.dto.cartitem;
+
+import jakarta.validation.constraints.Positive;
+
+public record CreateCartItemRequestDto(
+        @Positive(message = "Field bookId can't be null")
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/dto/cartitem/CreateCartItemRequestDto.java]:
```suggestion
        @Positive(message = "Field bookId can't be null")
        @NotNull
```

"Field bookId can't be null" wrong message for that annotation

File: src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+
+        cartItemRepository.findByShoppingCartIdAndBookId(shoppingCart.getId(),
+                        cartItemRequestDto.bookId())
+                .ifPresent(existingCartItem -> {
+                    throw new EntityNotFoundException("Book with Id: "
+                            + cartItemRequestDto.bookId()
+                            + " already exists in the shopping cart");
+                });
+
+        Book book = bookRepository.findById(cartItemRequestDto.bookId())
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/shopingcart/impl/ShoppingCartServiceImpl.java]: replace before  cartItemRepository.findByShoppingCartIdAndBookId to check if Book exists in DB, if not you don't call that method

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Comment by petro2061** [comment]
The method call order has been changed in addCartItemToShoppingCart in ShoppingCartServiceImpl class. Also chenged return type in private method in ShoppingCartController.

File: src/main/java/project/onlinebookstore/repository/cartitem/CartItemRepository.java [hunk 21]
```diff
@@ -0,0 +1,18 @@
+package project.onlinebookstore.repository.cartitem;
+
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+import project.onlinebookstore.model.CartItem;
+
+@Repository
```
Comment (MentorMateAcademy) [inline src/main/java/project/onlinebookstore/repository/cartitem/CartItemRepository.java]:
```suggestion
```

Comment (MentorMateAcademy) [reply src/main/java/project/onlinebookstore/repository/cartitem/CartItemRepository.java]: this anno redundant

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

**Comment by petro2061** [comment]
Deleted redundant annotation.

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**